pkg/ccl/streamingccl/streamclient/cockroach_sinkless_replication_client.go

```diff
@@ -124,8 +125,17 @@ func (m *sinklessReplicationClient) ConsumePartition(
 			}
 		}
 		if err := rows.Err(); err != nil {
-			errCh <- err
-			return
+			switch err {
+			case driver.ErrBadConn:
+				select {
+				case <-eventCh:
+					eventCh <- streamingccl.MakeGenerationEvent()
```

Comment:
I wonder if it makes sense for the ingestion processor to be aware of the concept of generations at all. Would it be simpler if we just swallowed `ErrBadConn` in the driver in the client which would hang the processor until its context was cancelled.

Also does `errors.Is(err, driver.ErrBadConn)` work here? I think that's the more standard way of checking for errors (see https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20190318_error_handling.md for more info).

Reply:
The processor would go to a different state If we swallow `ErrBadConn` on the client side, since there are no more values to be read from `eventCh`, consequently we'd stop reading from `cutoverCh`. I'm not sure if there's a better way of differentiating the error returned by the client.

Reply:
@pbardea  If I'm understanding correctly, the issue with swallowing the error is that we return from the sinkless client. The return then triggers the `eventCh` for the partition to be closed, which in turn causes the `merged` channel in the processor to be closed. Once the merged channel is closed, the proc drains and we're in a bit of a soup.

Thinking ahead to multi-partition where we probably want to do more than just wait for a cutover (ingest up until some time), a generation event would give us a place to trigger this custom logic? There might be a better way though

Reply:
Gotcha, my original thinking was that we wouldn't close the `eventCh` and the processors would be unaware that the client had disconnected, it would just hang waiting for its context to be canceled. (When the client sees the error it just reads from `<-ctx.Done()`.)

The tl;dr of below is that given that the coordinator doesn't yet have a way to send processors messages, this approach should be okay for now, but I think it's worth keeping an eye on potentially moving this to the coordinator down the line.

----

A bit more consideration to why it may be good to move to a scheme where the coordinator is the only one aware of "cutovers times" and "generation events":

Such a scheme would allow each ingestion processor to not need to worry about global state like generations and cutover time, but only to ingest data from the stream, and to watch for a "ingest until at least <x>" signal from the coordinator. (We don't have a way for the coordinator to provide a signal like that today).

The underlying concern is that on large clusters, we'll develop these many to 1 relationships that could put unnecessary pressure on some nodes. (We have this today with cutover time polling since all 100 nodes would poll the job record on avg 3 times a second and that would only increase as the polling interval is reduced).

Reply:
100% agree, filed https://github.com/cockroachdb/cockroach/issues/68475.